framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,51 +533,74 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
-	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	/* If no pending buffers ,then enqueue some buffers*/
+	if (pcm->enqueuedBytes==0) {
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
+		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
+#else
+		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
+#endif
+		{
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+			if (nbytes <= pcm->enqueuedBytes) {
+				break;
+			} else if (nbytes - pcm->enqueuedBytes < bufdesc.numbytes) {
+				bufdesc.numbytes = nbytes - pcm->enqueuedBytes;
+			}
+			bufdesc.u.pBuffer->nmaxbytes = bufdesc.numbytes;
+			bufdesc.u.pBuffer->nbytes = 0;
+			bufdesc.u.pBuffer->curbyte = 0;
+			bufdesc.u.pBuffer->flags = 0;
+
+			if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+				return oops(pcm, errno, "AUDIOIOC_ENQUEUEBUFFER ioctl failed");
+			}
+			pcm->enqueuedBytes += bufdesc.numbytes;
+		}
+		if ((!pcm->running) && pcm_start(pcm) < 0) {
+			return -errno;
+		}
 	}
 
-	/* Wait for deque message from kernel */
 	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
 	if (size != sizeof(msg)) {
 		/* Interrupted by a signal? What to do? */
 		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
 	}
 	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
 		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
+		pcm->enqueuedBytes -= apb->nmaxbytes;
+
 		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
```

Comment:
I think there can be a problem with this memcpy. 
Suppose we have 4 buffers of each 2000 bytes.
Suppose, the user calls readi with nbytes as below.

readi(100)
- we return 100 bytes
readi(4000)
- we enque 2 buffers of 2000 bytes each
- after first dequeue, suppose apb->nbytes = 2000 bytes. So, we return 2000 bytes
readi(100)
- when we dequeue here, we get apb->nbytes = 2000 bytes (due to previous enqueue)
- we copy 2000 bytes into a buffer meant for 100 bytes !!
- Not only this, we also loose track of the 100 bytes that user has requested on this present call to readi

Reply:
Yes, there is problem with memcpy . 
But i guess we wouldnt be loosing track of 100 bytes readi call because of those 2000 bytest first hundred would have been the one which user wanted.